packages/react/src/useActionForm.ts

```diff
@@ -0,0 +1,263 @@
+  useEffect(() => {
+    if (isReady && !hasSetInitialValues.current && defaultValues) {
+      hasSetInitialValues.current = true;
+      formHook.reset(defaultValues);
+    }
+  }, [isReady, defaultValues, formHook]);
+
+  // get the action runner to run on submit
+  const [actionResult, runAction] = useAction(action, { select: options?.select } as any);
+
+  const handleSubmissionError = useCallback(
+    (error: Error | FieldErrors<ActionFunc["variablesType"]>) => {
+      if ("executionErrors" in error) {
+        const executionErrors = (error as unknown as ErrorWrapper).executionErrors;
+        for (const executionError of executionErrors) {
+          if ("validationErrors" in executionError) {
+            for (const validationError of executionError.validationErrors) {
+              formHook.setError(`${action.modelApiIdentifier}.${validationError.apiIdentifier}` as any, {
+                message: validationError.message,
+              });
+            }
+          } else {
+            const codeToReplace = "code" in executionError ? `${executionError.code}: ` : "";
+            const message = executionError.message.replace(codeToReplace, "");
+
+            formHook.setError("root", { message });
+          }
+        }
+      }
+
+      options?.onError?.(error);
+    },
+    [action.modelApiIdentifier, formHook, options]
+  );
+
+  const submit = useCallback(
+    async (event?: React.BaseSyntheticEvent<object, any, any> | undefined) => {
+      let result: any;
+
+      formHook.clearErrors();
+
+      await handleSubmit(
+        async (data) => {
+          let modelInput = data[action.modelApiIdentifier];
+          if (modelInput) {
+            modelInput = { ...modelInput };
+            delete modelInput.__typename;
+            delete modelInput.id;
+            delete modelInput.createdAt;
+            delete modelInput.updatedAt;
+          }
+
+          let variables: Record<string, any> = {
+            id: existingRecord ? existingRecord.id : undefined,
+            ...data,
+            [action.modelApiIdentifier]: modelInput,
+          };
+
+          if (options?.send) {
+            const unmasked = variables;
+            variables = {};
+            for (const key of options.send) {
+              set(variables, key, get(unmasked, key));
+            }
+          }
+
+          options?.onSubmit?.();
+
+          result = await runAction(variables as any);
+          if (!result.error) {
+            options?.onSuccess?.(result.data);
+          }
+        },
+        (errors) => {
+          handleSubmissionError(errors);
+        }
+      )(event);
+
+      if (result?.error) {
+        handleSubmissionError(result.error);
+      }
+    },
+    [handleSubmit, formHook, handleSubmissionError, existingRecord, options, action.modelApiIdentifier, runAction]
+  );
+
+  const proxiedFormState = new Proxy(formState, {
```

Comment:
Is there a way to avoid this proxy? It seems kinda complicated and I find throws newer developers for a loop when you console.log it or see it in the debugger. Whats it for?

Reply:
the formState from useForm does some weird proxy stuff, from https://react-hook-form.com/docs/useform/formstate

```
formState is wrapped with a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) to improve render performance and skip extra logic if specific state is not subscribed to. Therefore make sure you invoke or read it before a render in order to enable the state update.```

I was initially thinking I would create my own state from the useForm state but I didn't want to mess with any optimizations they were doing

would be open to any suggestions on better ways to do this derived state though

Reply:
Ah, read a bit more about that and I think you're right, we need to keep it as a proxy to take advantage of their nice only-re-render-if-you-touch-the-thing stuff. Fancy!